controllers/certificate/certificate_controller.go

```diff
@@ -0,0 +1,170 @@
+// +kubebuilder:rbac:groups="apiextensions.k8s.io",resources=customresourcedefinitions,verbs=get;list;watch;create;update;patch;delete
+
+func (r *CertificateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.log.Info("Received reconciliation request", "name", req.Name)
+
+	certificateSecret := &corev1.Secret{}
+	err := r.Client.Get(ctx, req.NamespacedName, certificateSecret)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("Reconciling Webhook Secret", "name", certificateSecret.Name)
+
+	err = verifySecret(certificateSecret)
+	if err == nil {
+		r.log.Info("Secret certificate is still valid and does not need to be updated")
+	} else {
+		r.log.Info("Secret certificate is invalid", "verificationError", err.Error())
+		certificate, err := createNewSecret(ctx, r.Client, certificateSecret)
+		if err != nil {
+			return ctrl.Result{RequeueAfter: 1 * time.Minute}, err
+		}
+		r.log.Info("New certificate created", "validFrom", certificate.NotBefore, "validUntil", certificate.NotAfter)
+	}
+
+	return ctrl.Result{RequeueAfter: r.reconciliationInterval}, nil
+}
+
+func (r *CertificateReconciler) SetupWithManager(mgr ctrl.Manager, c controllers.RateLimiterConfig) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}).
+		WithEventFilter(predicate.NewPredicateFuncs(func(o client.Object) bool { return o.GetName() == secretName && o.GetNamespace() == secretNamespace })).
+		WithOptions(controller.Options{
+			RateLimiter: controllers.NewRateLimiter(c),
+		}).
+		Complete(r)
+}
+
+func createNewSecret(ctx context.Context, client ctrlclient.Client, secret *corev1.Secret) (*x509.Certificate, error) {
+	certificate, key, err := generateCertificate(serviceName, secret.Namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate certificate")
+	}
+
+	newSecret := buildSecret(secret.Name, secret.Namespace, certificate, key)
+	secret.Data = newSecret.Data
+
+	if err := client.Update(ctx, secret); err != nil {
+		return nil, errors.Wrap(err, "failed to update secret")
+	}
+
+	if err := updateCertificateInCRD(ctx, client, certificate); err != nil {
+		return nil, errors.Wrap(err, "failed to update certificate into CRD")
+	}
+
+	parsedCertificates, err := cert.ParseCertsPEM(certificate)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse certificate")
+	}
+
+	return parsedCertificates[0], nil
+}
+
+func generateCertificate(serviceName, namespace string) ([]byte, []byte, error) {
+	namespacedServiceName := strings.Join([]string{serviceName, namespace}, ".")
+	commonName := strings.Join([]string{namespacedServiceName, "svc"}, ".")
+	serviceHostname := fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, namespace)
+	altNames := []string{
+		commonName,
+		serviceName,
+		namespacedServiceName,
+		serviceHostname,
+	}
+	return generateSelfSignedCertificate(altNames[0], nil, altNames, maxAge)
```

Comment:
Maybe it's better to use the commonName as param, instead of relying on the order of the array?
```suggestion
	return generateSelfSignedCertificate(commonName, nil, altNames, maxAge)
```